sync/sync_head.go

```diff
@@ -186,7 +192,59 @@ func (s *Syncer[H]) verify(ctx context.Context, newHead H) (bool, error) {
 			"reason", verErr.Reason)
 	}
 
-	return verErr.SoftFailure, err
+	return false, err
+}
+
+// verifySkipping performs a bifurcated header verification process such that
+// it tries to find a header (or several headers if necessary)
+// between the networkHead and the subjectiveHead such that non-adjacent
+// (or in the worst case adjacent) verification passes and the networkHead
+// can be verified as a valid sync target against the syncer's subjectiveHead.
+//
+// When such candidates cannot be found [NewValidatorSetCantBeTrustedError] will be returned.
+func (s *Syncer[H]) verifySkipping(ctx context.Context, subjHead, networkHeader H) error {
+	subjHeight := subjHead.Height()
+
+	diff := networkHeader.Height() - subjHeight
+	if diff <= 0 {
+		panic(fmt.Sprintf("implementation bug:\n subjective head height %d, hash %X,\n network head height %d, hash %X",
+			subjHeight, subjHead.Hash(),
+			networkHeader.Height(), networkHeader.Hash(),
+		))
+	}
+
+	for diff > 1 {
+		candidateHeight := subjHeight + diff/2
+
+		candidateHeader, err := s.getter.GetByHeight(ctx, candidateHeight)
+		if err != nil {
+			return err
+		}
+
+		if err := header.Verify(subjHead, candidateHeader); err != nil {
+			// candidate failed, go deeper in 1st half.
+			diff = diff / 2
+			continue
+		}
+
+		// candidate was validated properly, update subjHead.
+		subjHead = candidateHeader
+		s.setSubjectiveHead(ctx, subjHead)
+
+		if err := header.Verify(subjHead, networkHeader); err == nil {
+			// network head validate properly, return success.
+			return nil
+		}
+
+		// new subjHead failed, go deeper in 2nd half.
+		subjHeight = subjHead.Height()
+		diff = networkHeader.Height() - subjHeight
+	}
+
+	s.metrics.failedValidationAgainstSubjHead(ctx, int64(networkHeader.Height()), networkHeader.Hash().String())
+	log.Warnw("sync: header validation against subjHead", "height", networkHeader.Height(), "hash", networkHeader.Hash().String())
+
+	return fmt.Errorf("sync: header validation against subjHead height:%d hash:%x", networkHeader.Height(), networkHeader.Hash().String())
```

Comment:
After thinking about this for a bit. I think we should return hard `VerifyError` to comply to the verification contract and log it on loc 187. Error handling for this case should go through the same code path as a regular verification failure.

Reply:
I vote for a separate error 'cause it's not just a bad header but also a try to verify it for sure. I'm talking more about logging here, not sure about hard `VerifyError`, no opinion on that.

We need a 3rd opinion here.

Reply:
I would say we need this error independent if we want to catch and handle it any different. But do we want that? My take is no. Currently treating it different would break logging